### PR TITLE
andreasbuhr-cppcoro: Conan 2.0 support

### LINKED ISSUE #16621: [request] andreasbuhr-cppcoro/latest
### Package Name/Version

andreasbuhr-cppcoro/latest

### Changelog

https://github.com/lewissbaker/cppcoro

### Context about the new update

The [version](https://conan.io/center/andreasbuhr-cppcoro) which is on conan.io fails to compile with clang.
For ``` conan install andreasbuhr-cppcoro/cci.20210113@_/_ --profile clang --build=missing ```, it says 
<details> 
<summary> Output </summary>
<br>

```
andreasbuhr-cppcoro/cci.20210113: WARN: Package binary is corrupted, removing: ce7925f094948675ea74d23df83a42941d85bfdd
Installing package: andreasbuhr-cppcoro/cci.20210113
Requirements
andreasbuhr-cppcoro/cci.20210113 from 'conancenter' - Cache
Packages
andreasbuhr-cppcoro/cci.20210113:ce7925f094948675ea74d23df83a42941d85bfdd - Build

Installing (downloading, building) binaries...
andreasbuhr-cppcoro/cci.20210113: WARN: Build folder is dirty, removing it: /home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd
andreasbuhr-cppcoro/cci.20210113: Copying sources to build folder
andreasbuhr-cppcoro/cci.20210113: Building your package in /home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd
andreasbuhr-cppcoro/cci.20210113: Generator cmake created conanbuildinfo.cmake
andreasbuhr-cppcoro/cci.20210113: Aggregating env generators
andreasbuhr-cppcoro/cci.20210113: Calling build()
-- The CXX compiler identification is Clang 15.0.7
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Adjusting fPIC flag (ON)
-- Conan: Compiler Clang>=8, checking major version 15
-- Conan: Checking correct version: 15
-- Conan: C++ stdlib: libstdc++11
C++ version not set. Defaulting to CMAKE_CXX_STANDARD=20
-- Performing Test _CXX_COROUTINES_SUPPORTS_MS_FLAG
-- Performing Test _CXX_COROUTINES_SUPPORTS_MS_FLAG - Failed
-- Performing Test _CXX_COROUTINES_SUPPORTS_MS_HEAPELIDE_FLAG
-- Performing Test _CXX_COROUTINES_SUPPORTS_MS_HEAPELIDE_FLAG - Failed
-- Performing Test _CXX_COROUTINES_SUPPORTS_TS_FLAG
-- Performing Test _CXX_COROUTINES_SUPPORTS_TS_FLAG - Success
-- Performing Test _CXX_COROUTINES_SUPPORTS_CORO_FLAG
-- Performing Test _CXX_COROUTINES_SUPPORTS_CORO_FLAG - Failed
-- Looking for C++ include coroutine
-- Looking for C++ include coroutine - found
-- Performing Test _CXX_COROUTINES_FINAL_HEADER_COMPILES
-- Performing Test _CXX_COROUTINES_FINAL_HEADER_COMPILES - Success
-- Performing Test CXX_COROUTINES_NO_AWAIT_NEEDED
-- Performing Test CXX_COROUTINES_NO_AWAIT_NEEDED - Success
-- Configuring done
-- Generating done
CMake Warning:
Manually-specified variables were not used by the project:

CMAKE_EXPORT_NO_PACKAGE_REGISTRY
CMAKE_INSTALL_BINDIR
CMAKE_INSTALL_DATAROOTDIR
CMAKE_INSTALL_INCLUDEDIR
CMAKE_INSTALL_LIBDIR
CMAKE_INSTALL_LIBEXECDIR
CMAKE_INSTALL_OLDINCLUDEDIR
CMAKE_INSTALL_SBINDIR


-- Build files have been written to: /home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd
[ 15%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/async_mutex.cpp.o
[ 26%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/cancellation_token.cpp.o
[ 15%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/cancellation_state.cpp.o
[ 26%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/async_manual_reset_event.cpp.o
[ 26%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/async_auto_reset_event.cpp.o
[ 31%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/cancellation_source.cpp.o
[ 36%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/cancellation_registration.cpp.o
[ 47%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ip_address.cpp.o
[ 47%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/lightweight_manual_reset_event.cpp.o
[ 52%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ipv4_address.cpp.o
[ 57%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ip_endpoint.cpp.o
[ 63%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ipv4_endpoint.cpp.o
[ 68%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ipv6_address.cpp.o
[ 73%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/ipv6_endpoint.cpp.o
[ 78%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/static_thread_pool.cpp.o
[ 84%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/auto_reset_event.cpp.o
[ 89%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/spin_wait.cpp.o
[ 94%] Building CXX object source_subfolder/lib/CMakeFiles/cppcoro.dir/spin_mutex.cpp.o
/home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd/source_subfolder/lib/static_thread_pool.cpp:670:23: error: no member named 'exchange' in namespace 'std'; did you mean '__exchange'?
auto* next = std::exchange(tail->m_next, head);
~~~~~^~~~~~~~
__exchange
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/move.h:157:5: note: '__exchange' declared here
__exchange(_Tp& __obj, _Up&& __new_val)
^
/home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd/source_subfolder/lib/static_thread_pool.cpp:671:17: error: no member named 'exchange' in namespace 'std'; did you mean '__exchange'?
head = std::exchange(tail, next);
~~~~~^~~~~~~~
__exchange
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/move.h:157:5: note: '__exchange' declared here
__exchange(_Tp& __obj, _Up&& __new_val)
^
2 errors generated.
gmake[2]: *** [source_subfolder/lib/CMakeFiles/cppcoro.dir/build.make:272: source_subfolder/lib/CMakeFiles/cppcoro.dir/static_thread_pool.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:113: source_subfolder/lib/CMakeFiles/cppcoro.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
andreasbuhr-cppcoro/cci.20210113:
andreasbuhr-cppcoro/cci.20210113: ERROR: Package 'ce7925f094948675ea74d23df83a42941d85bfdd' build failed
andreasbuhr-cppcoro/cci.20210113: WARN: Build folder /home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd
ERROR: andreasbuhr-cppcoro/cci.20210113: Error in build() method, line 82
cmake.build()
ConanException: Error 2 while executing cmake --build '/home/akshit/.conan/data/andreasbuhr-cppcoro/cci.20210113/_/_/build/ce7925f094948675ea74d23df83a42941d85bfdd' '--' '-j12'
```
</details>

<details>
<summary> Configuration: </summary>

```
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=clang
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
os_build=Linux
[options]
[build_requires]
[env]
CC=/usr/bin/clang
CXX=/usr/bin/clang++
```
</details>

This error is [fixed](https://github.com/andreasbuhr/cppcoro/blob/10bbcdbf2be3ad3aa56febcf4c7662d771460a99/lib/static_thread_pool.cpp#L15) in the updated version of the library.

Thanks

### LINKED ISSUE #20564: Library andreasbuhr-cppcoro is not compatible with Conan v2
This issue comes from the Clion plugin

### PR DESCRIPTION
**Pull request by Ahajha**
Specify library name and version:  **andreasbuhr-cppcoro/all**

I noticed the recipe previously had a single component defined, which seemed unnecessary, so I removed it. If anyone might know why that was there I can re-add it.

Closes #20564 
Closes #16621 
---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:20]: Missing `package_type = "library"` attribute

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:60]: Is this still the case? https://godbolt.org/z/GndEK54GY compiles for me

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:100]: This flag is no longer valid in newer clang versions afaik, should check its version beforhand

**Comment by RubenRBS** [comment]
Hi @Ahajha thanks a lot for your PR, we really appreciate it! :)
The windows error comes from the fact that a check for C++17 at least should be done, else CCI will set it to 14 in windows and it won't compile

Just checked that there's a similar PR (I thought my first review didn't go thru!) for this recipe, maybe we can work to bring them both together with their changes?

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:5]: Good catch, I checked the min version before I made that change.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:60]: Might be difficult for me to test, but I can give it a shot. I get very confused when it comes to switching around compilers haha

**Comment by Ahajha** [comment]
@RubenRBS I didn't even notice there was another PR for this until I had already pushed it, but it seems this one will be a bit easier since it has a few more months of changes/fixes. I can definitely incorporate the rest of that branch's changes.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:60]: It looks like the other PR removed it as well, I'll take that at face value and just go ahead and remove it.

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: All the clang/libstdc++ checks failed, I'll see if I can get them working, otherwise I'll re-add this error.

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: Testing locally, it doesn't work until clang 14.

**Comment by Ahajha** [comment]
Getting a C5231 on MSVC in the test package, which has almost no documentation online, I can't visually tell what's causing it, and I can't reproduce it on my machine. So for now trying to remove 20210113 and see if it's fixed upstream.

**Comment by Ahajha** [comment]
Not quite ready yet:
1. Need to figure out why cci.20210113 is failing on Windows, maybe we either drop support or bump it to a slightly later commit. Regardless I think having the latest is a good idea.
2. Add system deps for Windows
3. Minor test package cleanup (v1 should defer to v2)

**Comment by Ahajha** [comment]
Looking at the changes between cci.20210113 and the latest, it's mostly very small bugfixes (including the windows build issue), so I imagine it won't cause any breakages for people to upgrade. Thus, I think we can just leave only the latest version. If someone needs a previous version we can revisit.

With that note, this should be ready for another review once CI passes.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: No idea how I'm going to test apple-clang, I only have apple-clang 14 and it works, so I'm going to leave it at that.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:110]: Per comment above, apple-clang 14 (and 15, after upgrading) works as-is, so I'm going to leave it.

**Comment by Ahajha** [comment]
@RubenRBS This is ready for review (not sure if there's another process to signaling ready-ness)

**Comment by RubenRBS** [comment]
Pinging works, thanks :) (If not, we usually come around the PR eventually after a few days)

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:82]: FYI (but no need to change and force a recompilation just because of this): This can always be set to ON, it's a noop in every other context afaik :)

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:111]: > (no compiler explorer :( )

Damn, let me try to see if there's something I can do on that front, I'm also a main collaborator there!

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: What if we check C++20 in a branch for gcc etc, and else check C++17 for clang & msvc?

**Review by RubenRBS — COMMENTED** [review]
Thanks! Somall questions, but looks nice

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: Will test for you on Monday

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: We're already checking 17 across the board, but I can see if GCC with C++17 fails.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:82]: Good to know, if I need to change something else I'll fix this as well.

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: GCC 10+ in C++17 mode works, so checking for 20 would potentially exclude some valid configurations. So actually the comment seems to apply to all compilers, they all work in C++17 mode. If anything, I think mentioning the C++ standard is confusing, really all we're checking for is the coroutines TS, as the error message says. The C++17 check above is sufficient for the C++ standard check.

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: So perhaps we can say something like "In addition to C++17, cppcoro requires Coroutine TS support".

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: @RubenRBS Can you please give this another look? I think this is the last actionable comment (or if you wanted to test some clang configurations).

Comment (uilianries) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:57]: @Ahajha Our maintainer, Rubén, is on vacation, but will return on next Monday 😄

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: Well, would have done so if I didn't forget I would be out on vacation! I'm doing so now, thanks a lot for your patience :)

Comment (RubenRBS) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: Approving, thanks a lot for doublechecking!

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py]: Did you intend to approve the PR? If so, you haven't yet @RubenRBS

**Comment by Ahajha** [comment]
@RubenRBS If you don't have any other comments, can you please approve? I would like to get this merged.

**Comment by Ahajha** [comment]
@RubenRBS Can you please take another look at this PR to see if there's any more required changes?

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (RubenRBS) [inline recipes/andreasbuhr-cppcoro/all/conanfile.py:44]: This will be an issue for compilers that only have coroutines support for C++20 enabled. This is the case of clang >= 16, where only having `-std=c++20` enables coroutines. With this check, CCI will only try to build for C++ 17. Maybe a solution is to force C++20 as a min required version if the other flag checks have not succeeded?

**Review by RubenRBS — COMMENTED** [review]
Thanks. I have a concern with Clang >= 16. I've proposed a solution, and I'm interested in hearing what you think about it :)

File: recipes/andreasbuhr-cppcoro/all/conanfile.py
Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:44]: Maybe we could just update this value to be
```return 20 if self.settings.compiler == "clang" and self.settings.compiler.version >= 16 else 17```?

Comment (Ahajha) [reply recipes/andreasbuhr-cppcoro/all/conanfile.py:44]: There's also a funny edge case that I just ran into: When using libstdc++ with clang, it always required C++20. With libc++, it allows C++17 actually through clang 16, but fails with 17. I'll double check all my tests, but assuming that's correct I'll update the checks.

**Comment by Ahajha** [comment]
@RubenRBS A couple things I found while doing some more extensive clang testing:
1. It seems that the -fcoroutines-ts flag is never needed, at least as far down as clang11. Just to be on the safe side, for versions older than that, I kept the flag.
2. If using newer libstdc++ versions (I have libstdc++ 13, not sure if earlier ones will cause it) with clang 14 specifically, the build may fail for an unrelated reason. I don't know if there's a setting to test this, but I have it outputting a warning for now if it detects libstdc++ and clang 14.
3. libstdc++ + clang always requires c++20
4. clang allows c++17 mode up to _and including_ clang 16, clang 17 is the first one to require c++20.

**Comment by Ahajha** [comment]
CI didn't run? Going to close + reopen

**Comment by Ahajha** [comment]
@RubenRBS I think this is ready for another look (I don't see the Conan v2 pipeline though, odd)

**Comment by Ahajha** [comment]
@RubenRBS Can you take another look at this one?

**Review by RubenRBS — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**
